pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,348 @@
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore the root as well
+					notFound = true
+					break
+				}
+				vs := delegate.Spec.(*networking.VirtualService)
+				merged := mergeHttpRoute(route, vs.Http)
+				mergedRoutes = append(mergedRoutes, merged...)
+			} else {
+				mergedRoutes = append(mergedRoutes, route)
+			}
+		}
+		if !notFound {
+			rootVs.Http = mergedRoutes
+			out = append(out, root)
+		}
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHttpRoute(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		// suppose there are N1 match conditions in root, N2 match conditions in delegate
+		// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
+		merged, conflict := mergeHTTPMatchRequests(root.Match, subRoute.Match)
+		if conflict {
+			log.Debugf("HTTPMatchRequests conflict: root root %s, delegate root %s", root.Name, subRoute.Name)
+			continue
+		}
+		subRoute.Match = merged
+
+		if subRoute.Name == "" {
+			subRoute.Name = root.Name
+		}
+		if subRoute.Rewrite == nil {
+			subRoute.Rewrite = root.Rewrite
+		}
+		if subRoute.Timeout == nil {
+			subRoute.Timeout = root.Timeout
+		}
+		if subRoute.Retries == nil {
+			subRoute.Retries = root.Retries
+		}
+		if subRoute.Fault == nil {
+			subRoute.Fault = root.Fault
+		}
+		if subRoute.Mirror == nil {
+			subRoute.Mirror = root.Mirror
+		}
+		if subRoute.MirrorPercentage == nil {
+			subRoute.MirrorPercentage = root.MirrorPercentage
+		}
+		if subRoute.CorsPolicy == nil {
+			subRoute.CorsPolicy = root.CorsPolicy
+		}
+
+		// merge headers
+		subRoute.Headers = mergeHeaders(root.Headers, subRoute.Headers)
+		out = append(out, subRoute)
+	}
+	return out
+}
+
+// return merged match conditions if not conflicts
+func mergeHTTPMatchRequests(root, delegate []*networking.HTTPMatchRequest) (out []*networking.HTTPMatchRequest, conflict bool) {
+	if len(root) == 0 {
+		return delegate, false
+	}
+
+	if len(delegate) == 0 {
+		return root, false
+	}
+
+	// each HTTPMatchRequest of delegate must find a superset in root.
+	// otherwise it conflicts
+	for _, subMatch := range delegate {
+		foundMatch := false
+		for _, rootMatch := range root {
+			if hasConflict(rootMatch, subMatch) {
+				continue
+			}
+			// merge HTTPMatchRequest
+			out = append(out, mergeHTTPMatchRequest(rootMatch, subMatch))
+			foundMatch = true
+		}
+		if !foundMatch {
+			return nil, true
+		}
+	}
+	if len(out) == 0 {
+		conflict = true
+	}
+	return
+}
+
+func mergeHTTPMatchRequest(root, delegate *networking.HTTPMatchRequest) *networking.HTTPMatchRequest {
```

Comment:
how can we make this maintainable when new fields are added?

Reply:
fuzz test is a good way

Reply:
done